### PR TITLE
KAFKA-16564: Apply `Xlint` to java code in core module

### PR DESCRIPTION
**Pull request by m1a2st**
Jira: https://issues.apache.org/jira/browse/KAFKA-16564

We should apply the same `xlint` to java code in core module before we complete the code migration.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -119,8 +119,14 @@ ext {
   addParametersForTests = { name, options ->
     // -parameters generates arguments with parameter names in TestInfo#getDisplayName.
     // ref: https://github.com/junit-team/junit5/blob/4c0dddad1b96d4a20e92a2cd583954643ac56ac0/junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTest.java#L161-L164
-    if (name == "compileTestJava" || name == "compileTestScala")
+    if (name == "compileTestJava" || name == "compileTestScala") {
       options.compilerArgs << "-parameters"
+    } else if (name == "compileJava") {
```
Comment (chia7712) [inline build.gradle]: please change `compileJava` to `compileScala` to add the args to "scala compiler"

**Comment by chia7712** [comment]
@m1a2st could you please rebase code and fix conflicts?

**Comment by m1a2st** [comment]
I have been fix it, Thanks for your reminder.

**Comment by chia7712** [comment]
@m1a2st could you please rebase code and fix conflicts?

File: build.gradle [hunk 2]
```diff
@@ -119,8 +119,17 @@ ext {
...
+    if (name == "compileTestJava" || name == "compileTestScala") {
       options.compilerArgs << "-parameters"
+    } else if (name == "compileScala") {
+      options.compilerArgs << "-Xlint:all"
+      options.compilerArgs << "-encoding" << "UTF-8"
+      options.compilerArgs << "-Xlint:-rawtypes"
+      options.compilerArgs << "-Xlint:-serial"
+      options.compilerArgs << "-Xlint:-try"
+      options.compilerArgs << "-Werror"
+      options.compilerArgs << "-Xlint:-options"
```
Comment (chia7712) [inline build.gradle]: why we need this option?

Comment (m1a2st) [reply build.gradle]: The `-Xlint:-options` is "Warns about the issues relating to use of command line options" .  [doc](https://docs.oracle.com/en/java/javase/17/docs/specs/man/javac.html)
I think we don't need this option, I will remove it.

File: build.gradle [hunk 3]
```diff
@@ -119,8 +119,16 @@ ext {
   addParametersForTests = { name, options ->
     // -parameters generates arguments with parameter names in TestInfo#getDisplayName.
     // ref: https://github.com/junit-team/junit5/blob/4c0dddad1b96d4a20e92a2cd583954643ac56ac0/junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTest.java#L161-L164
-    if (name == "compileTestJava" || name == "compileTestScala")
+    if (name == "compileTestJava" || name == "compileTestScala") {
       options.compilerArgs << "-parameters"
+    } else if (name == "compileScala") {
```
Comment (chia7712) [inline build.gradle]: Could you please config both scala and java? for example:
```
    if (name == "compileTestJava" || name == "compileTestScala") {
      options.compilerArgs << "-parameters"
    } else if (name == "compileJava" || name == "compileScala") {
      options.compilerArgs << "-Xlint:all"
      if (!project.path.startsWith(":connect") && !project.path.startsWith(":storage"))
        options.compilerArgs << "-Xlint:-rawtypes"
      options.compilerArgs << "-encoding" << "UTF-8"
      options.compilerArgs << "-Xlint:-rawtypes"
      options.compilerArgs << "-Xlint:-serial"
      options.compilerArgs << "-Xlint:-try"
      options.compilerArgs << "-Werror"
      options.compilerArgs += ["--release", String.valueOf(minJavaVersion)]
    }
```

File: build.gradle [hunk 4]
```diff
@@ -119,8 +119,16 @@ ext {
   addParametersForTests = { name, options ->
     // -parameters generates arguments with parameter names in TestInfo#getDisplayName.
     // ref: https://github.com/junit-team/junit5/blob/4c0dddad1b96d4a20e92a2cd583954643ac56ac0/junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTest.java#L161-L164
-    if (name == "compileTestJava" || name == "compileTestScala")
```
Comment (chia7712) [inline build.gradle:119]: Additionally, `addParametersForTests` -> `configureJavaCompiler`

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for your patch

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/SynchronizationTest.java [hunk 5]
```diff
@@ -461,9 +461,7 @@ private static ThreadFactory threadFactoryWithNamedThreads(String threadPrefix)
         return r -> {
             // This is essentially Executors.defaultThreadFactory except with
             // custom thread names so in order to filter by thread names when debugging
-            SecurityManager s = System.getSecurityManager();
-            Thread t = new Thread((s != null) ? s.getThreadGroup() :
-                Thread.currentThread().getThreadGroup(), r,
+            Thread t = new Thread(Thread.currentThread().getThreadGroup(), r,
```
Comment (m1a2st) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/SynchronizationTest.java]: Java 17 is deprecated the SecurityManager, see [doc](https://openjdk.org/jeps/411)

**Comment by chia7712** [comment]
@m1a2st please fix build error

File: build.gradle [hunk 6]
```diff
@@ -113,11 +113,22 @@ ext {
...
+    if (name == "compileTestJava" || name == "compileTestScala") {
       options.compilerArgs << "-parameters"
+    } else if (name == "compileJava" || name == "compileScala") {
+      options.compilerArgs << "-Xlint:all"
+      if (!project.path.startsWith(":connect") && !project.path.startsWith(":storage"))
+        options.compilerArgs << "-Xlint:-rawtypes"
+      options.compilerArgs << "-encoding" << "UTF-8"
+      options.compilerArgs << "-Xlint:-rawtypes"
```
Comment (chia7712) [inline build.gradle:126]: duplicate config?

File: clients/src/test/java/org/apache/kafka/common/security/ssl/CommonNameLoggingTrustManagerFactoryWrapperTest.java [hunk 7]
```diff
@@ -122,7 +122,7 @@ void testNeverExpiringX509Certificate() throws Exception {
                     wrappedCert.hasUnsupportedCriticalExtension());
             // We have just generated a valid test certificate, it should still be valid now
             assertEquals(cert.getBasicConstraints(), wrappedCert.getBasicConstraints());
-            assertEquals(cert.getIssuerDN(), wrappedCert.getIssuerDN());
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/security/ssl/CommonNameLoggingTrustManagerFactoryWrapperTest.java]: why we need those changes?

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for this patch

File: build.gradle (hunk 6 above)
Comment (m1a2st) [reply build.gradle:126]: `configureJavaCompiler` also use in the `tasks.withType(JavaCompile)` and `subprojects`, it is dulpicate in the `tasks.withType(JavaCompile)`code block so I remove it which in `tasks.withType(JavaCompile)`

File: clients/src/test/java/org/apache/kafka/common/security/ssl/CommonNameLoggingTrustManagerFactoryWrapperTest.java (hunk 7 above)
Comment (m1a2st) [reply clients/src/test/java/org/apache/kafka/common/security/ssl/CommonNameLoggingTrustManagerFactoryWrapperTest.java]: it seens this deprecated method won't cause build error, I will revert it

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/SynchronizationTest.java (hunk 5 above)
Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/SynchronizationTest.java]: Could you please use suppression to fix it? we can go back to dig in it later

**Review by chia7712 — APPROVED** [review]

File: build.gradle [hunk 8]
```diff
@@ -771,7 +773,7 @@ subprojects {
     scalaCompileOptions.additionalParameters += ["-release", String.valueOf(minJavaVersion)]
     options.compilerArgs += ["--release", String.valueOf(minJavaVersion)]
```
Comment (chia7712) [inline build.gradle]: Could you please remove this duplicate config?

**Pull request merged**